src/Ace/Halogen/Component.purs

```diff
@@ -1,75 +1,213 @@
 module Ace.Halogen.Component
   ( aceComponent
   , aceConstructor
-  , module Ace.Halogen.Component.Query
-  , module Ace.Halogen.Component.State
+  , AceEffects()
+  , Autocomplete(..)
+  , AceState(..)
+  , AceQuery(..)
+  , CompleteFn()
   ) where
 
 import Prelude
 
+import Ace.Editor as Editor
+import Ace.Ext.LanguageTools as LanguageTools
+import Ace.Ext.LanguageTools.Completer as Completer
+import Ace.Types
 import Control.Monad (when)
+import Control.Monad.Aff (Aff(), runAff)
+import Control.Monad.Aff.AVar (AVAR())
+import Control.Monad.Eff (Eff())
 import Control.Monad.Eff.Class (MonadEff)
-
-import Data.Maybe (Maybe(..))
-
-import Halogen (Natural(), Component(), ComponentHTML(), ComponentDSL(), SlotConstructor(..), component, action, liftEff', liftH, modify, gets)
+import Control.Monad.Eff.Random (random, RANDOM())
+import Control.Monad.Eff.Ref (Ref(), REF(), readRef, writeRef, modifyRef)
+import DOM (DOM())
+import DOM.HTML.Types (HTMLElement())
+import Data.Date (nowEpochMilliseconds, Now())
+import Data.Foldable (traverse_)
+import Data.Maybe (Maybe(..), maybe)
+import Data.StrMap (StrMap())
+import Data.StrMap as Sm
+import Data.Time (Milliseconds(..))
+import Halogen
 import Halogen.HTML.Indexed as H
 import Halogen.HTML.Properties.Indexed as P
 
-import Ace.Types (ACE(), Editor())
-import Ace.Editor as Editor
-import Ace.Halogen.Component.Query
-import Ace.Halogen.Component.State
+----------------------------------------------------------------------
+-- Foreign knots and effectful functions
+----------------------------------------------------------------------
+-- | Effectful knot of autocomplete functions. It's needed because
+-- | `languageTools.addCompleter` is global and adds completer to
+-- | all editors
+foreign import completeFns :: forall e. Ref (StrMap (CompleteFn e))
```

Comment:
I think there's a potential memory leak here as we only ever insert into to `completeFns`, we should probably delete from it when finalizing the components or disabling autocomplete (assuming we unify `SetAutocomplete` and `SetCompleteFn`).